pynumaflow/tests/function/test_async_server.py

```diff
@@ -0,0 +1,242 @@
+import asyncio
+import logging
+import threading
+import unittest
+from typing import AsyncIterable
+
+import grpc
+
+from google.protobuf import timestamp_pb2 as _timestamp_pb2
+from grpc.aio._server import Server
+
+from pynumaflow import setup_logging
+from pynumaflow._constants import DATUM_KEY, WIN_START_TIME, WIN_END_TIME
+from pynumaflow.function import Messages, Message, Datum, Metadata, UserDefinedFunctionServicer
+from pynumaflow.function.generated import udfunction_pb2, udfunction_pb2_grpc
+from pynumaflow.tests.function.test_server import (
+    map_handler,
+    mock_event_time,
+    mock_watermark,
+    mock_message,
+    mock_interval_window_start,
+    mock_interval_window_end,
+)
+
+LOGGER = setup_logging(__name__)
+
+
+async def async_reduce_handler(key: str, datums: AsyncIterable[Datum], md: Metadata) -> Messages:
+    interval_window = md.interval_window
+    counter = 0
+    async for _ in datums:
+        counter += 1
+    msg = (
+        f"counter:{counter} interval_window_start:{interval_window.start} "
+        f"interval_window_end:{interval_window.end}"
+    )
+
+    return Messages(Message.to_vtx(key, str.encode(msg)))
+
+
+def request_generator(count, request, resetkey: bool = False):
+    for i in range(count):
+        if resetkey:
+            request.key = f"key-{i}"
+        yield request
+
+
+def start_reduce_streaming_request() -> (Datum, tuple):
+    event_time_timestamp = _timestamp_pb2.Timestamp()
+    event_time_timestamp.FromDatetime(dt=mock_event_time())
+    watermark_timestamp = _timestamp_pb2.Timestamp()
+    watermark_timestamp.FromDatetime(dt=mock_watermark())
+
+    request = udfunction_pb2.Datum(
+        value=mock_message(),
+        event_time=udfunction_pb2.EventTime(event_time=event_time_timestamp),
+        watermark=udfunction_pb2.Watermark(watermark=watermark_timestamp),
+    )
+
+    metadata = (
+        (DATUM_KEY, "test"),
+        (WIN_START_TIME, f"{mock_interval_window_start()}"),
+        (WIN_END_TIME, f"{mock_interval_window_end()}"),
+    )
+    return request, metadata
+
+
+_s: Server = None
+_channel = grpc.insecure_channel("localhost:50051")
+_loop = None
+
+
+def startup_callable(loop):
+    asyncio.set_event_loop(loop)
+    loop.run_forever()
+
+
+async def start_server():
+    server = grpc.aio.server()
+    udfs = UserDefinedFunctionServicer(
+        reduce_handler=async_reduce_handler,
+        map_handler=map_handler,
+    )
+    udfunction_pb2_grpc.add_UserDefinedFunctionServicer_to_server(udfs, server)
+    listen_addr = "[::]:50051"
+    server.add_insecure_port(listen_addr)
+    logging.info("Starting server on %s", listen_addr)
+    global _s
+    _s = server
+    await server.start()
+    await server.wait_for_termination()
+
+
+class TestAsyncServer(unittest.TestCase):
```

Comment:
I think if we were to test the async methods separately from /functions/server.py, yes we will need them
however, since we will be making non-async calls to gRPC endpoint, this will suffice for this test case.